File: src/figurestead/presentation.py
"""Opt-in presentation poses that compose existing scientific geometry."""

from __future__ import annotations

from dataclasses import dataclass
from copy import deepcopy
from typing import Mapping, Any

import matplotlib.patheffects as pe
import numpy as np

from .themes import Theme


@dataclass(frozen=True)
class ScientificPose:
    panel_surface: bool = True
    frame: bool = True
    curve: str = "monotone"
    line_width: float = 2.35
    marker_scale: float = 1.22
    series_markers: tuple[str, ...] = ("o", "s")
    legend_location: str = "lower right"


@dataclass(frozen=True)
class FocusAnnotation:
    x: float
    y: float
    label: str
    offset: tuple[float, float] = (42.0, -22.0)


@dataclass(frozen=True)
class EvidenceFocusAnnotation:
    mark_id: str
    label: str
    offset: tuple[float, float] = (42.0, -22.0)


SCIENTIFIC_POSE = ScientificPose()


def apply_scientific_pose(contract: Mapping[str, Any], *, focus: Mapping[str, FocusAnnotation | EvidenceFocusAnnotation] | None = None) -> dict[str, Any]:
    """Clone a portable figure and add the same opt-in pose used by the web renderer."""
    result = deepcopy(dict(contract))
    panels = result.get("panels")
    if not isinstance(panels, list) or not panels:
        raise ValueError("contract.panels must be a non-empty list")
    focus = focus or {}
    known = {str(panel.get("id")) for panel in panels}
    unknown = sorted(set(focus) - known)
    if unknown:
        raise ValueError(f"unknown focus panels: {', '.join(unknown)}")
    for panel in panels:
        presentation = {
            "panelSurface": True, "frame": True, "legend": "auto", "markerScale": 1.22,
            **({"curve": "monotone", "lineWidth": 2.35, "seriesMarkers": ["ring", "square"]} if panel.get("renderer") == "line" else {}),
            **dict(panel.get("presentation", {})),
        }
        panel["presentation"] = presentation
        if panel.get("renderer") == "line":
            panel["encoding"] = {**dict(panel.get("encoding", {})), "interpolation": "monotone"}
        item = focus.get(str(panel.get("id")))
        if item is not None:
            annotation = {"type": "focus", "label": item.label, "dx": item.offset[0], "dy": -item.offset[1]}
            if isinstance(item, EvidenceFocusAnnotation):
                if not item.mark_id.strip():
                    raise ValueError("EvidenceFocusAnnotation.mark_id must be non-empty")
                annotation["anchorId"] = item.mark_id
            else:
                annotation.update({"x": item.x, "y": item.y})
            panel["annotations"] = [*panel.get("annotations", []), annotation]
    return result


def apply_evidence_pose(contract: Mapping[str, Any], *, focus: Mapping[str, EvidenceFocusAnnotation]) -> dict[str, Any]:
    """Apply a pose whose callouts can only target compiled evidence mark IDs."""
    if any(not isinstance(item, EvidenceFocusAnnotation) for item in focus.values()):
        raise TypeError("apply_evidence_pose focus values must be EvidenceFocusAnnotation instances")
    return apply_scientific_pose(contract, focus=focus)


def resolve_pose(pose: ScientificPose | str | None) -> ScientificPose | None:
    if pose is None:
        return None
    if pose == "scientific":
        return SCIENTIFIC_POSE
    if isinstance(pose, ScientificPose):
        return pose
    raise ValueError("pose must be None, 'scientific', or ScientificPose")


def monotone_curve(x, y, *, samples_per_segment: int = 18):
    """Densify strictly increasing samples with shape-preserving cubic Hermite curves."""
    x, y = np.asarray(x, dtype=float), np.asarray(y, dtype=float)
    if len(x) < 2 or len(x) != len(y) or np.any(np.diff(x) <= 0):
        return x, y
    width = np.diff(x)
    delta = np.diff(y) / width
    tangent = np.empty_like(x)
    tangent[0], tangent[-1] = delta[0], delta[-1]
    if len(x) > 2:
        tangent[1:-1] = (delta[:-1] + delta[1:]) / 2
    for index, slope in enumerate(delta):
        if slope == 0:
            tangent[index:index + 2] = 0
            continue
        left, right = tangent[index] / slope, tangent[index + 1] / slope
        length = left * left + right * right
        if length > 9:
            scale = 3 / np.sqrt(length)
            tangent[index], tangent[index + 1] = scale * left * slope, scale * right * slope
    curve_x, curve_y = [], []
    for index in range(len(x) - 1):
        t = np.linspace(0, 1, samples_per_segment, endpoint=False)
        h00, h10 = 2 * t**3 - 3 * t**2 + 1, t**3 - 2 * t**2 + t
        h01, h11 = -2 * t**3 + 3 * t**2, t**3 - t**2
        curve_x.extend(x[index] + width[index] * t)
        curve_y.extend(h00 * y[index] + h10 * width[index] * tangent[index] + h01 * y[index + 1] + h11 * width[index] * tangent[index + 1])
    curve_x.append(x[-1]); curve_y.append(y[-1])
    return np.asarray(curve_x), np.asarray(curve_y)


def draw_focus_annotation(ax, focus: FocusAnnotation, theme: Theme):
    fill = theme.summary_core
    edge = theme.series_edges[0] if theme.series_edges else (theme.primary_edge or theme.field)
    ax.scatter([focus.x], [focus.y], s=210, facecolors="none", edgecolors=theme.primary,
               linewidths=5.0, alpha=0.13, zorder=7)
    ax.scatter([focus.x], [focus.y], s=115, facecolors=fill, edgecolors=edge,
               linewidths=2.0, zorder=8)
    annotation = ax.annotate(
        focus.label, xy=(focus.x, focus.y), xytext=focus.offset, textcoords="offset points",
        color=fill, fontsize=8.3, fontfamily="DejaVu Sans Mono", fontweight="bold",
        arrowprops={"arrowstyle": "-", "color": fill, "linewidth": 1.25},
        ha="left" if focus.offset[0] >= 0 else "right", va="center", zorder=9,
    )
    annotation.set_path_effects([pe.Stroke(linewidth=2.4, foreground=theme.field), pe.Normal()])
    return annotation

File: web/src/presentation.js
import { cloneValue, windowProgress } from "./schema.js";
import { clamp01, smooth } from "./marks.js";

export const SCIENTIFIC_POSE = Object.freeze({
  panelSurface: true,
  frame: true,
  curve: "monotone",
  legend: "auto",
  lineWidth: 2.35,
  markerScale: 1.22,
  seriesMarkers: Object.freeze(["ring", "square"]),
});

export function applyScientificPose(input, options = {}) {
  const result = cloneValue(input);
  if (!Array.isArray(result?.panels) || !result.panels.length) throw new TypeError("applyScientificPose requires a v0.4 contract with one or more panels");
  const panelIds = new Set(result.panels.map((panel) => panel.id));
  result.panels = result.panels.map((panel) => {
    const presentation = {
      panelSurface: true,
      frame: true,
      legend: options.legend ?? "auto",
      markerScale: options.markerScale ?? SCIENTIFIC_POSE.markerScale,
      ...(panel.renderer === "line" ? {
        curve: options.curve ?? "monotone",
        lineWidth: options.lineWidth ?? SCIENTIFIC_POSE.lineWidth,
        seriesMarkers: options.seriesMarkers ?? ["ring", "square"],
      } : {}),
      ...(panel.presentation ?? {}),
    };
    return {
      ...panel,
      presentation,
      ...(panel.renderer === "line" ? { encoding: { ...(panel.encoding ?? {}), interpolation: options.curve ?? panel.presentation?.curve ?? "monotone" } } : {}),
    };
  });
  for (const focus of options.focus ?? []) {
    if (!panelIds.has(focus.panelId)) throw new TypeError(`unknown focus panel ${focus.panelId}`);
    const panel = result.panels.find((item) => item.id === focus.panelId);
    panel.annotations = [...(panel.annotations ?? []), {
      type: "focus", ...(focus.anchorId ? { anchorId: focus.anchorId } : { x: focus.x, y: focus.y, space: focus.space }), label: focus.label,
      dx: focus.dx ?? 68, dy: focus.dy ?? 28,
    }];
  }
  return result;
}

export function applyEvidencePose(input, options = {}) {
  const focus = options.focus ?? [];
  if (!Array.isArray(focus)) throw new TypeError("options.focus must be an array");
  focus.forEach((item, index) => {
    if (typeof item?.panelId !== "string" || !item.panelId.trim()) throw new TypeError(`focus[${index}].panelId must identify a panel`);
    if (typeof item.anchorId !== "string" || !item.anchorId.trim()) throw new TypeError(`focus[${index}].anchorId must identify a compiled evidence mark`);
    if (typeof item.label !== "string" || !item.label.trim()) throw new TypeError(`focus[${index}].label must be a non-empty string`);
  });
  return applyScientificPose(input, { ...options, focus });
}

export function drawPanelSurface(context, { contract, layout }) {
  if (!contract.presentation?.panelSurface) return;
  const { plot } = layout;
  context.save();
  context.fillStyle = contract.theme.panel;
  context.fillRect(plot.left, plot.top, plot.right - plot.left, plot.bottom - plot.top);
  if (contract.presentation.frame) {
    context.strokeStyle = contract.theme.spine;
    context.globalAlpha = 0.48;
    context.lineWidth = Math.max(0.6, 0.75 * layout.scale);
    context.strokeRect(plot.left, plot.top, plot.right - plot.left, plot.bottom - plot.top);
  }
  context.restore();
}

function position(annotation, scales, plot) {
  if (annotation.space === "plot") {
    return {
      x: plot.left + clamp01(annotation.x) * (plot.right - plot.left),
      y: plot.bottom - clamp01(annotation.y) * (plot.bottom - plot.top),
    };
  }
  if (typeof scales?.x !== "function" || typeof scales?.y !== "function") return null;
  const x = scales.x(annotation.x), y = scales.y(annotation.y);
  return Number.isFinite(x) && Number.isFinite(y) ? { x, y } : null;
}

export function drawPresentationAnnotations(context, { contract, layout, scales, progress }) {
  const focus = contract.annotations.filter((item) => item?.type === "focus");
  if (!focus.length) return;
  const reveal = smooth(windowProgress(progress, contract.timeline.summaryCompiles));
  if (reveal <= 0) return;
  const { plot } = layout, theme = contract.theme;
  focus.forEach((annotation) => {
    const point = position(annotation, scales, plot);
    if (!point || typeof annotation.label !== "string" || !annotation.label.trim()) return;
    const dx = Number.isFinite(annotation.dx) ? annotation.dx * layout.scale : 68 * layout.scale;
    const dy = Number.isFinite(annotation.dy) ? annotation.dy * layout.scale : 28 * layout.scale;
    const rawX = point.x + dx, rawY = point.y + dy;
    const labelX = Math.max(plot.left + 26 * layout.scale, Math.min(plot.right - 26 * layout.scale, rawX));
    const labelY = Math.max(plot.top + 22 * layout.scale, Math.min(plot.bottom - 20 * layout.scale, rawY));
    const radius = Math.max(6.5, 8.4 * layout.scale) * reveal;
    const fill = theme.summaryCore;
    const edge = theme.seriesEdges?.[0] ?? theme.primaryEdge ?? theme.field;
    context.save();
    context.globalAlpha = reveal;
    context.strokeStyle = theme.primary;
    context.lineWidth = Math.max(3, radius * 0.62);
    context.globalAlpha = 0.13 * reveal;
    context.beginPath(); context.arc(point.x, point.y, radius * 1.45, 0, Math.PI * 2); context.stroke();
    context.globalAlpha = 0.92 * reveal;
    context.strokeStyle = fill;
    context.lineWidth = Math.max(1, 1.25 * layout.scale);
    context.beginPath(); context.moveTo(point.x + radius * 0.7, point.y + radius * 0.55); context.lineTo(labelX - 8 * layout.scale, labelY - 5 * layout.scale); context.stroke();
    context.fillStyle = fill;
    context.strokeStyle = edge;
    context.lineWidth = Math.max(1.4, 1.9 * layout.scale);
    context.beginPath(); context.arc(point.x, point.y, radius, 0, Math.PI * 2); context.fill(); context.stroke();
    context.font = `600 ${Math.max(9, layout.font.legend * 1.06)}px ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, monospace`;
    context.textAlign = dx < 0 ? "right" : "left";
    context.textBaseline = "middle";
    context.strokeStyle = theme.field;
    context.lineWidth = Math.max(1.6, 2.2 * layout.scale);
    context.strokeText(annotation.label, labelX, labelY);
    context.fillStyle = fill;
    context.fillText(annotation.label, labelX, labelY);
    context.restore();
  });
}
